File: src/main/scala/im/tox/toktok/app/main/friends/FriendsRecyclerAdapter.scala
package im.tox.toktok.app.main.friends

import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.ViewHolder
import android.view.{ LayoutInflater, View, ViewGroup }
import android.widget.RelativeLayout
import im.tox.toktok.TypedResource._
import im.tox.toktok.app.domain.Friend
import im.tox.toktok.{ R, TR }

abstract class FriendsRecyclerAdapter(
    friends: java.util.List[Friend],
    expandOnClick: FriendItemClicks
) extends RecyclerView.Adapter[FriendsRecyclerViewHolder] {

  private var expandedItem = -1

  def onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): FriendsRecyclerViewHolder = {
    val itemView = LayoutInflater.from(viewGroup.getContext).inflate(TR.layout.fragment_home_friends_item, viewGroup, false)

    new FriendsRecyclerViewHolder(itemView, new FriendItemExpand {
      override def onClick(): Unit = {
        val holder = itemView.getTag.asInstanceOf[ViewHolder]

        if (expandedItem == holder.getLayoutPosition) {
          notifyItemChanged(expandedItem)
          expandedItem = -1
        } else {
          if (expandedItem >= 0) {
            notifyItemChanged(expandedItem)
          }

          expandedItem = holder.getLayoutPosition
          notifyItemChanged(expandedItem)
        }
      }
    })
  }

  def onBindViewHolder(viewHolder: FriendsRecyclerViewHolder, position: Int): Unit = {
    val item = friends.get(position)
    viewHolder.mUserName.setText(item.userName)
    viewHolder.mUserImage.setImageResource(item.photoReference)
    viewHolder.mUserImage.setClickable(true)
    viewHolder.mUserImage.setOnClickListener(new View.OnClickListener {
      override def onClick(v: View): Unit = {
        expandOnClick.startOverLayFriend(position)
      }
    })

    if (position == expandedItem) {
      viewHolder.mExpandedArea.setVisibility(View.VISIBLE)
      viewHolder.mBase.setElevation(10)
      viewHolder.mBase.setBackgroundResource(R.drawable.cardboard_ripple)

      viewHolder.mCallButton.setOnClickListener(new View.OnClickListener {
        override def onClick(v: View): Unit = {
          expandOnClick.startCall(position)
        }
      })

      viewHolder.mMessageButton.setOnClickListener(new View.OnClickListener {
        override def onClick(v: View): Unit = {
          expandOnClick.startMessage(position)
        }
      })
    } else {
      viewHolder.mExpandedArea.setVisibility(View.GONE)
      viewHolder.mBase.setElevation(0)
      viewHolder.mBase.setBackgroundResource(R.drawable.background_ripple)

      viewHolder.mCallButton.setOnClickListener(null)
      viewHolder.mMessageButton.setOnClickListener(null)
    }

  }

  def getItemCount: Int = {
    friends.size()
  }

  def getItemPosition(i: Int): String = {
    friends.get(i).userName
  }

  def getItem(i: Int): Friend = {
    friends.get(i)
  }

}

final class FriendsRecyclerViewHolder(
    itemView: RelativeLayout,
    itemOnClick: FriendItemExpand
) extends RecyclerView.ViewHolder(itemView)
  with View.OnClickListener {

  val mBase = itemView.findView(TR.home_friends_base)
  val mUserInfo = itemView.findView(TR.home_item_info)
  val mUserName = itemView.findView(TR.home_friends_name)
  val mUserImage = itemView.findView(TR.home_friends_img)
  val mExpandedArea = itemView.findView(TR.home_friends_expanded_area)
  val mCallButton = itemView.findView(TR.home_friends_call)
  val mMessageButton = itemView.findView(TR.home_friends_message)

  mBase.setTag(this)
  itemView.setOnClickListener(this)

  def onClick(view: View): Unit = {
    itemOnClick.onClick()
  }

}

trait FriendItemExpand {
  def onClick(): Unit
}

trait FriendItemClicks {
  def startOverLayFriend(layoutPosition: Int): Unit
  def startCall(layoutPosition: Int): Unit
  def startMessage(layoutPosition: Int): Unit
}

File: sources/app/src/main/java/im/tox/toktok/app/main/friends/FriendsRecyclerAdapter.java
package im.tox.toktok.app.main.friends;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.RecyclerView.ViewHolder;

import java.util.List;

import de.hdodenhof.circleimageview.CircleImageView;
import im.tox.toktok.R;
import im.tox.toktok.app.domain.Friend;

abstract class FriendsRecyclerAdapter extends RecyclerView.Adapter<FriendsRecyclerViewHolder> {

    private final List<Friend> friends;
    private final FriendItemClicks expandOnClick;

    private int expandedItem = -1;

    FriendsRecyclerAdapter(
            List<Friend> friends,
            FriendItemClicks expandOnClick
    ) {
        this.friends = friends;
        this.expandOnClick = expandOnClick;
    }

    @NonNull
    @Override
    public FriendsRecyclerViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        final RelativeLayout itemView = (RelativeLayout) LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.fragment_home_friends_item, viewGroup, false);

        return new FriendsRecyclerViewHolder(itemView, () -> {
            ViewHolder holder = (ViewHolder) itemView.getTag();

            if (expandedItem == holder.getLayoutPosition()) {
                notifyItemChanged(expandedItem);
                expandedItem = -1;
            } else {
                if (expandedItem >= 0) {
                    notifyItemChanged(expandedItem);
                }

                expandedItem = holder.getLayoutPosition();
                notifyItemChanged(expandedItem);
            }
        });
    }

    @Override
    public void onBindViewHolder(@NonNull final FriendsRecyclerViewHolder viewHolder, int position) {
        Friend item = friends.get(position);
        viewHolder.mUserName.setText(item.userName);
        viewHolder.mUserImage.setImageResource(item.photoReference);
        viewHolder.mUserImage.setClickable(true);
        viewHolder.mUserImage.setOnClickListener(v -> expandOnClick.startOverLayFriend(viewHolder.getAdapterPosition()));

        if (position == expandedItem) {
            viewHolder.mExpandedArea.setVisibility(View.VISIBLE);
            viewHolder.mBase.setElevation(10);
            viewHolder.mBase.setBackgroundResource(R.drawable.cardboard_ripple);

            viewHolder.mCallButton.setOnClickListener(v -> expandOnClick.startCall(viewHolder.getAdapterPosition()));

            viewHolder.mMessageButton.setOnClickListener(v -> expandOnClick.startMessage(viewHolder.getAdapterPosition()));
        } else {
            viewHolder.mExpandedArea.setVisibility(View.GONE);
            viewHolder.mBase.setElevation(0);
            viewHolder.mBase.setBackgroundResource(R.drawable.background_ripple);

            viewHolder.mCallButton.setOnClickListener(null);
            viewHolder.mMessageButton.setOnClickListener(null);
        }

    }

    @Override
    public int getItemCount() {
        return friends.size();
    }

    String getItemPosition(int i) {
        return friends.get(i).userName;
    }

    public Friend getItem(int i) {
        return friends.get(i);
    }

}

final class FriendsRecyclerViewHolder extends RecyclerView.ViewHolder
        implements View.OnClickListener {

    private final FriendItemExpand itemOnClick;
    final RelativeLayout mBase;
    final RelativeLayout mUserInfo;
    final TextView mUserName;
    final CircleImageView mUserImage;
    final LinearLayout mExpandedArea;
    final TextView mCallButton;
    final TextView mMessageButton;

    FriendsRecyclerViewHolder(
            @NonNull RelativeLayout itemView,
            FriendItemExpand itemOnClick
    ) {
        super(itemView);
        this.itemOnClick = itemOnClick;
        mBase = itemView.findViewById(R.id.home_friends_base);
        mUserInfo = itemView.findViewById(R.id.home_item_info);
        mUserName = itemView.findViewById(R.id.home_friends_name);
        mUserImage = itemView.findViewById(R.id.home_friends_img);
        mExpandedArea = itemView.findViewById(R.id.home_friends_expanded_area);
        mCallButton = itemView.findViewById(R.id.home_friends_call);
        mMessageButton = itemView.findViewById(R.id.home_friends_message);

        mBase.setTag(this);
        itemView.setOnClickListener(this);
    }


    @Override
    public void onClick(View view) {
        itemOnClick.onClick();
    }

}

